Ensure reports are json_encodable. (Fix silently failing reports with Malformed UTF8 in stacktrace)

YasserB94 opened the pull request:
## Why?

Whilst creating a WordPress plugin to send exceptions to Flare. I noticed reports sent, when I invoked them from the wp-cli, they would silently fail.

In 1.0 I had a middleware that would just remove the whole stack trace to resolve this. I spoke to @rubenvanassche about this in a mail-thread a couple weeks ago. 

After some digging, I noticed that an error would be thrown if I added `JSON_THROW_ON_ERROR` to the ReportTrimmer here:
https://github.com/spatie/flare-client-php/blob/3c280c074498215ef9a44ae4a4df3dd8d4b4a37a/src/Truncation/ReportTrimmer.php#L39-L43

it would throw (`JsonException: Malformed UTF-8 characters, possibly incorrectly encoded`),
further down the road I noticed that, even should I pave the road for the report to reach the API. Either curl would fail, or the API wouldn't like it.

**TLDR:** Non UTF-8 characters, from binaries end up in  the stack trace, silently making the report fail.

## Changes in this PR 

**Disclaimer:** I tried to do this right. Although I figured I'd just create a PR and iterate on feedback instead of spending years attempting to wrap some code with a golden bow. 

I added a Support class that checks whether an array can be json_encoded. If it can't, it recurses through the whole payload. Replacing any entry that cannot be json_encoded. 

For the replacement, I opted to add the error message, the amount of bytes omitted, and a Prefix so it's clear that this was replaced by the client internally  


## Additional Ideas/caveats
- Add a config option to enable/disable this behaviour
- Add an option for early checkout, checking if the whole $payload is fixed upon each replacement and stopping recursion early
- There is a lot going on in Flare (Thank you Spatie!), I feel like there is a better place to put this code. If anyone shares the same itch, I would love to learn
- I think currently a `resource` would be replaced by null in `json_encode`? Though now it would be replaced by a NON ENCODABLE message.
- Not sure if I should add a test about this behaviour in ReportTest as well? (Though it may be some repetition asserting that Malformed UTF is replaced.)

## Art 
After adding these changes, I was able to successfully see my exception/report in flare. 
<img width="1003" height="749" alt="stacktrace_wp-cli" src="https://github.com/user-attachments/assets/1c9fe104-4df4-4519-b093-cc22f8b773dc" />

YasserB94 committed:
ReportSanitizer| Class to ensure a $payload is json encodable

YasserB94 committed:
Report | Sanitize payload to be json_encodable in toArray

YasserB94 committed:
Fix: Phpstan typing errors in ReportSanitizer

YasserB94 commented:
I noticed the following issue in the phpstan workflow ([71cae65](https://github.com/spatie/flare-client-php/pull/42/commits/71cae651eadc5b183c9144f5e8f10a7e986926aa)) output:
```
Error: Method Spatie\FlareClient\Report::toArray() should return array<string, mixed> but returns array<int|string, mixed>.
Error: Process completed with exit code 1.
```

This is fixed in [3660995](https://github.com/spatie/flare-client-php/pull/42/commits/3660995a2df066821bd6b1f1e028aaa012f5fd14). Though brought up a typing caveat... 

As in the following method:
```php
 /**
     * @template K of array-key
     *
     * @param  array<K,mixed>  $input
     * @return array<K,mixed>
     */
    protected static function replaceNonEncodableEntries(array $input): array
    {
        foreach ($input as $key => $value) {
            try {
                json_encode($input[$key], JSON_THROW_ON_ERROR);
            } catch (JsonException $e) {
                if (is_string($value)) {
                    $input[$key] = sprintf('[%s: CORRUPT DATA]: %d bytes | ERR: %s', self::REPLACED_ENTRY_PREFIX, strlen($value), $e->getMessage());
                } elseif (is_array($value)) {
                    $input[$key] = self::replaceNonEncodableEntries($value);
                } else {
                    $input[$key] = sprintf('[%s: NON ENCODABLE]: type %s', self::REPLACED_ENTRY_PREFIX, gettype($value));
                }
            }
        }

        return $input;
    }
    ```
    Should the value be anything other than a string, it would also become a string. Not sure if this arrises an issue? And my `else` statement on the replacement part should ensure the type of the value remains intact?

rubenvanassche commented:
Hi @YasserB94,

Sorry for the late response to your PR we've been quite busy with the whole performance monitoring launch. While I like the idea, you're the first one to have this problem and think it is not that useful to spend computer cycles on making a JSON string from an array to throw it away again.

So what I propose, is that we keep the sanitizer as is, add it to the CurlSender & GuzzleSender with an option to enable sanitizing the payload but disabled by default.

That way we transform to JSON when we need it and immediately also do it for traces.

Last question, does this happen a lot in Wordpress that you've got malformed UTF-8? Haven't had such an issue with Laravel in the last 7 years or so.

Kind regards,

Ruben

YasserB94 committed:
Merge branch 'main' into fix/prevent-json-encoding-crashes

YasserB94 committed:
Refactor: Sanitize payloads optionally in Senders instead of Report

- Added shared sanitization behaviour into an AbstractSender class
  - Injected the PayloadSanitizer here
  - Optionally call it in the prepare payload method when configured
  - Used the existing config array to take in the option
  - added the logic from ReportSanitizer into JsonEncodableSanitizer
- Bound the sanitizer in the container

YasserB94 committed:
Fix JsonEncodableSanitizer - Don't sanitize if the payload has no errors

rubenvanassche commented:
Hi @YasserB94,

Sorry have been busy with working on the new clients and kinda forgot this PR, it is indeed quite heavy for a few users probably having this problem. Since it can be solved with a custom middleware I think that is (for now) the way to go. As soon as we get more complaints about these kinds of errors we can off course again revaluate if this code should be merged.

Thanks!

YasserB94 commented:
> Hi @YasserB94,
> 
> Sorry have been busy with working on the new clients and kinda forgot this PR, it is indeed quite heavy for a few users probably having this problem. Since it can be solved with a custom middleware I think that is (for now) the way to go. As soon as we get more complaints about these kinds of errors we can off course again revaluate if this code should be merged.
> 
> Thanks!

No worries, thanks for your response! 

Should you ever think of re-evaluating this, please feel free to ping me. 

Yet, as time passes I feel like there must be a more elegant solution to this problem should it ever become one big enough to solve.
